[GDAL provider] Use same underlying GDAL dataset for clone() (fixes #16006)

rouault opened the pull request:
This will improve efficiency of GDAL block cache during raster rendering,
since the cached blocks are only valid during GDALDataset instance lifetime.

rouault committed:
[GDAL provider] Use same underlying GDAL dataset for clone() (fixes #16006)

This will improve efficiency of GDAL block cache during raster rendering,
since the cached blocks are only valid during GDALDataset instance lifetime.

nyalldawson commented on src/providers/gdal/qgsgdalprovider.h:
```diff
@@ -158,6 +163,15 @@ class QgsGdalProvider : public QgsRasterDataProvider, QgsGdalProviderBase
+    QgsGdalProvider( const QgsGdalProvider &other );
+
+    // reference counter to know how many main and cloned provider instances are linked
+    int *mpRefCounter = nullptr;
```
Is a  QAtomicInt more appropriate here?

rouault replied:
should be a QAtomicInt* since we want the same counter to be shared by all clones. Anyway I already needed a mutex, so the use of the counter within the mutex is safe

nyalldawson replied:
Ah, of course!

+1 to merge from me

nyalldawson reviewed (commented):
Nice changes! I'm keen to see what impact this has on rendering speed

nirvn commented:
This might also help reduce some slowness introduced by the map canvas previews rendering.

rouault commented:
I'm still working on this as the above approach introduces lock contention for slow datasets (such as those using REST APIs), which makes the UI unresponsive

rouault committed:
[GDAL provider] Improve UI responsiveness affected by 3b2b98b012fdfda14761914f01318d079aeaa409

The consequence of sharing the same GDAL object and protecting them under mutex is that
instanciating a new pipeline for job preview background can actually be blocked for
a long time, and thus causing the UI to be totally non-responsive

So change the strategy to use a global cache of GDAL datasets that can be reused by
cloned GDAL providers, while still maintaining the number of opened datasets to
a reasonable amount.
